src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadGroup.java

```diff
@@ -428,24 +433,47 @@ public void saveProgram(File f) throws IOException {
             return;
         }
 
-        try {
-            RiverWriter w = new RiverWriter(tmpFile, execution.getOwner());
+        for (int i = 0; i < MAX_RETRIES; ++i) {
             try {
-                w.writeObject(this);
+                RiverWriter w = new RiverWriter(tmpFile, execution.getOwner());
+                try {
+                    w.writeObject(this);
+                } finally {
+                    w.close();
+                }
+                Files.move(tmpFile.toPath(), f.toPath(), StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING);
+                LOGGER.log(FINE, "program state saved");
+                break;
+            } catch (RuntimeException e) {
```

Comment:
No reported case is of anything other than an `IOException` so I would not want to retry in the face of other errors.